Revert pendingManifest cache and ask sub for manifest

wizeguyy opened the pull request:
The original idea of caching pending manifests was flawed. This PR reverts that commit and implements a better solution: when `InsertChain()` is called for a block whose manifest we don't know, we ask our sub for the manifest and rewrite the block body with the correct manifest.

wizeguyy commented:
Note, there's a few maintenance commits:
2x bugfixes -- small tangential fixes to related modules
1x SQUASH commit -- this will get squashed down when `geth-fork/cross-chain-txs` gets rebased next, but it needs to merge before the reversion commit in this PR

I can move these to independent PRs if anyone would like, but they seemed pertinent enough to this PR to include them as dependent commits. If I move them, then those other PR(s) will need to merge before this PR 🤷

kiltsonfire commented on core/slice.go:
```diff
@@ -374,6 +374,14 @@ func (sl *Slice) GetPendingHeader() (*types.Header, error) {
+func (sl *Slice) GetSubManifest(blockHash common.Hash) (types.BlockManifest, error) {
+	header := sl.hc.GetHeaderByHash(blockHash)
+	if header == nil {
+		return nil, fmt.Errorf("block not found")
```
typically we use errors.New("block not found") don't know if this will break error catching

wizeguyy replied:
fixed: [6ab0fd7](https://github.com/dominant-strategies/go-quai/pull/463/commits/6ab0fd74272c6ea3aa2edccd1134620b4cd3143a)

kiltsonfire commented on core/headerchain.go:
```diff
@@ -92,24 +92,23 @@ func NewHeaderChain(db ethdb.Database, engine consensus.Engine, chainConfig *par
 	manifest := types.BlockManifest{}
-	if h.NumberU64() == 0 {
+	// Terminate the search on coincidence or genesis
+	if hc.engine.HasCoincidentDifficulty(h) || h.NumberU64() == 0 {
```
This would probably be better to use a hash check rather than a number check.

wizeguyy replied:
Fixed: [5ec5fe9](https://github.com/dominant-strategies/go-quai/pull/463/commits/5ec5fe9f33328f38b04c285c27c24c65bb02a90c)

wizeguyy commented:
@kiltsonfire I've addressed your comments. I will perform one final rebase to squash commits down (as indicated in commit messages) once everyone is OK with the patch set.

alanorwick commented on core/core.go:
```diff
@@ -40,6 +41,21 @@ func (c *Core) InsertChain(blocks types.Blocks) (int, error) {
+		// If we just mined this block, its possible the subordinate manifest in our
+		// block body is incorrect. If so, ask our sub for the correct manifest,
+		// update and rewrite the correct body.
+		if block.ManifestHash() != block.SubManifest().Hash() {
```
How are we validating manifests for application? Here it just feels like a simple check of what's in the body compared to the header. Are we doing an ask every time to validate the subordinate manifest if we have that subClient?

wizeguyy replied:
When our dom node starts the append process, he passes in the manifest hash stored in his header. We check here that the manifest hash the dom gave us matches the hash of the manifest we compute locally for this block:
https://github.com/dominant-strategies/go-quai/blob/5ec5fe9f33328f38b04c285c27c24c65bb02a90c/core/headerchain.go#L140
